### PR TITLE
Create new BridgeStorageProvider getter/setter methods for bridge transactions index

### PR DESCRIPTION
**Pull request by nathanieliov**

### CODE REVIEW

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+        );
+        Mockito.reset(repository);
+
+        // Try to set a new sigHash that no exists in memory not in the index.
+        Sha256Hash newSigHash = PegTestUtils.createHash(7);
+        provider.setBridgeBtcTxSigHash(newSigHash);
+
+        // Verify the new sigHash is not persisted into the index when save has not been called.
+        verify(repository, never()).addStorageBytes(
+            any(),
```
Comment (jeremy-then) [inline rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java:219]: Is there any way to use the actual type for each of the parameters here? And in the other tests.

**Review by jeremy-then — APPROVED** [review]
Over all looks good to me. Just a minor optional comment in case is possible.

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java (hunk 1 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java:219]: `any()` is not a type. It refers to the value. It can be any value but it must be a value of the type expected for the method. I'm using any here because I expect this method never(`never()`) to get called, so I don't care about the value passing to the method. I care about this method not being called on this test.

Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java:219]: Oh sorry, I misunderstood your point. I went to the documentation and now I get you what you mean.

image

According to the `any()` matches everything, including null and `varargs`. On the other hand `any(Class<T> type)` excludes null. For this case, the one I want to is `any()`  because I can guarantee the method is never called no matter what value is passed.

Comment (jeremy-then) [reply rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java:219]: Ok.

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+            any(),
+            any()
+        );
+
+        // Now let's persist pending to save entries
+        provider.save();
+        verify(repository, times(1)).addStorageBytes(
```
Comment (jeremy-then) [inline rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java:226]: Here I think it could go like:

```java
verify(repository, times(1)).addStorageBytes(
            any(RskAddress.class),
            any(DataWord.class),
            any(byte[].class)
        );
```

What do you think?

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+            any(),
+            any(),
+            any()
+        );
+        Mockito.reset(repository);
+
+        // Now let's persist added sigHashes
+        provider.save();
+
+        verify(repository, times(2)).addStorageBytes(
```
Comment (jeremy-then) [inline rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java:368]: And here too.

```suggestion
        verify(repository, times(1)).addStorageBytes(
            any(RskAddress.class),
            any(DataWord.class),
            any(byte[].class)
        );
```

Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java:370]: Done

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java (hunk 2 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java:228]: Done

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+        Assertions.assertTrue(shouldFoundSigHash);
+        verify(repository, times(1)).getStorageBytes(
+            PrecompiledContracts.BRIDGE_ADDR,
+            BRIDGE_BTC_TX_SIG_HASH_KEY.getCompoundKey("-", sigHash.toString())
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("null_sigHash_parameters")
+    void setBridgeBtcTxSigHash_null_sigHash(boolean isRskip379Active, Sha256Hash sigHash) throws IOException {
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java:203]: Same here, sighash is always null

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+        verify(repository, never()).addStorageBytes(
+            any(),
+            any(),
+            any()
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("non_null_sigHash_parameters")
+    void hasBtcTxSigHash_non_null_sigHash(boolean isRskip379Active, Sha256Hash sigHash) {
```
[the remark this answers is not in the document]
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java:86]: Let me know what you think

File: rskj-core/src/main/java/co/rsk/peg/BridgeStorageIndexKey.java [hunk 6]
```diff
@@ -39,6 +39,9 @@ public enum BridgeStorageIndexKey {
     BTC_BLOCK_HEIGHT("btcBlockHeight"),
     FAST_BRIDGE_HASH_USED_IN_BTC_TX("fastBridgeHashUsedInBtcTx"),
     FAST_BRIDGE_FEDERATION_INFORMATION("fastBridgeFederationInformation"),
+
+    // Bridge btc tx sig hash index key
```
[the remark this answers is not in the document]
Comment (nathanieliov) [reply rskj-core/src/main/java/co/rsk/peg/BridgeStorageIndexKey.java]: Removed

File: rskj-core/src/main/java/co/rsk/peg/BridgeStorageIndexKey.java [hunk 7]
```diff
@@ -39,6 +39,9 @@ public enum BridgeStorageIndexKey {
...
+    BRIDGE_BTC_TX_SIG_HASH_KEY("bridgeBtcTxSigHash"),
```
[the remark this answers is not in the document]
Comment (nathanieliov) [reply rskj-core/src/main/java/co/rsk/peg/BridgeStorageIndexKey.java]: Updated.

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+class BridgeStorageProviderBridgeTxsIndexTests {
+
+    private static final String DUPLICATED_INSERTION_ERROR_MESSAGE = "Given bridge btc tx sigHash %s already exists in the index. Index entries are considered unique.";
+    private static final byte TRUE_VALUE = (byte) 1;
+    private final BridgeConstants bridgeConstantsRegtest = BridgeRegTestConstants.getInstance();
```
[the remark this answers is not in the document]
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java]: Updated.

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+            Arguments.of(false, Sha256Hash.ZERO_HASH),
+            Arguments.of(false, PegTestUtils.createHash(10)),
+            Arguments.of(true, Sha256Hash.ZERO_HASH),
+            Arguments.of(true, PegTestUtils.createHash(20))
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("null_sigHash_parameters")
+    void hasBtcTxSigHash_null_sigHash(boolean isRskip379Active, Sha256Hash sigHash) {
```
[the remark this answers is not in the document]
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java]: Updated.

File: rskj-core/src/main/java/co/rsk/peg/BridgeStorageProvider.java [hunk 10]
```diff
@@ -110,6 +110,8 @@ public class BridgeStorageProvider {
 
     private Long nextPegoutHeight;
 
+    private Sha256Hash bridgeBtcTxSigHashToSave;
```
Comment (kelvinator07) [inline rskj-core/src/main/java/co/rsk/peg/BridgeStorageProvider.java:113]:
```suggestion
    private Sha256Hash bridgeBtcTxSigHash;
```

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+    private final BridgeConstants bridgeConstants = BridgeMainNetConstants.getInstance();
+
+    private static Stream<Arguments> null_sigHash_parameters() {
+        return Stream.of(
+            Arguments.of(false),
+            Arguments.of(true)
+        );
+    }
+
+    private static Stream<Arguments> non_null_sigHash_parameters() {
```
Comment (kelvinator07) [inline rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java:41]:
```suggestion
    private static Stream<Arguments> valid_sigHash_parameters() {
```

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+        verify(repository, never()).getStorageBytes(
+            any(),
+            any()
+        );
+
+        verify(repository, never()).addStorageBytes(
+            any(),
+            any(),
+            any()
+        );
```
Comment (kelvinator07) [inline rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java:82]: The any() Method Be More Explicit By Passing The Data Type Its Expecting.. any(RskAddress.class)

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java [hunk 13]
```diff
@@ -0,0 +1,354 @@
...
+    }
+
+    @ParameterizedTest
+    @MethodSource("non_null_sigHash_parameters")
+    void hasBtcTxSigHash_non_null_sigHash(boolean isRskip379HardForkActive, Sha256Hash sigHash) {
```
Comment (kelvinator07) [inline rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java:87]:
```suggestion
    @MethodSource("valid_sigHash_parameters")
    void hasBtcTxSigHash_valid_sigHash(boolean isRskip379HardForkActive, Sha256Hash sigHash) {
```

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java [hunk 14]
```diff
@@ -0,0 +1,354 @@
...
+        // Arrange
+        ActivationConfig.ForBlock activations = isRskip379HardForkActive?
```
Comment (kelvinator07) [inline rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java:89]:
```suggestion
        ActivationConfig.ForBlock activations = isRskip379HardForkActive ?
```
You Can Run **_option + command + L_** On New Codes/Files To Properly Format The Code

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java [hunk 15]
```diff
@@ -0,0 +1,354 @@
...
+            BRIDGE_BTC_TX_SIG_HASH.getCompoundKey("-", sigHash.toString())
+        );
+        Mockito.reset(repository);
+
+        // Let's create a stub for the just saved sigHash
+        when(repository.getStorageBytes(PrecompiledContracts.BRIDGE_ADDR, BRIDGE_BTC_TX_SIG_HASH.getCompoundKey("-", sigHash.toString())))
+            .thenReturn(new byte[]{TRUE_VALUE});
+
+        // Check if saved sigHash exists
+        boolean shouldFoundSigHash = provider.hasBridgeBtcTxSigHash(sigHash);
```
Comment (kelvinator07) [inline rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java:194]:
```suggestion
        boolean shouldFindSigHash = provider.hasBridgeBtcTxSigHash(sigHash);
```

File: rskj-core/src/main/java/co/rsk/peg/BridgeStorageIndexKey.java [hunk 16]
```diff
@@ -39,6 +39,8 @@ public enum BridgeStorageIndexKey {
     BTC_BLOCK_HEIGHT("btcBlockHeight"),
     FAST_BRIDGE_HASH_USED_IN_BTC_TX("fastBridgeHashUsedInBtcTx"),
     FAST_BRIDGE_FEDERATION_INFORMATION("fastBridgeFederationInformation"),
+
+    BRIDGE_BTC_TX_SIG_HASH("bridgeBtcTxSigHash"),
     ;
 
     private String key;
```
Comment (kelvinator07) [inline rskj-core/src/main/java/co/rsk/peg/BridgeStorageIndexKey.java:46]:
```suggestion
    private final String key;
```

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java [hunk 17]
```diff
@@ -0,0 +1,354 @@
...
+        // Let's set the sigHash and then call hasBtcTxSigHash, it should return false.
+        provider.setBridgeBtcTxSigHash(sigHash);
+        verify(repository, times(1)).getStorageBytes(
+            PrecompiledContracts.BRIDGE_ADDR,
+            BRIDGE_BTC_TX_SIG_HASH.getCompoundKey("-", sigHash.toString())
+        );
+        // reset calls counter
+        Mockito.reset(repository);
+
+        boolean sigHashShouldNotExists = provider.hasBridgeBtcTxSigHash(sigHash);
```
Comment (kelvinator07) [inline rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java:168]:
```suggestion
        boolean sigHashShouldNotExist = provider.hasBridgeBtcTxSigHash(sigHash);
```

File: rskj-core/src/main/java/co/rsk/peg/BridgeStorageProvider.java [hunk 18]
```diff
@@ -923,6 +925,47 @@ protected int getReleaseRequestQueueSize() throws IOException {
...
+        }
+
+        byte[] data = repository.getStorageBytes(
+            contractAddress,
+            getStorageKeyForBridgeBtcTxSigHash(sigHash)
+        );
+
+        return data != null &&
+           data.length == 1 &&
+           data[0] == TRUE_VALUE;
```
Comment (kelvinator07) [inline rskj-core/src/main/java/co/rsk/peg/BridgeStorageProvider.java:940]: Can We Have Tests For This?
When Data Is Null, When Data Length Is More Than 1, When Data[0] Is FALSE_VALUE; 0

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java (hunk 12 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java:82]: Any() with explicit type exclude nulls. In this case, I'm testing using null values, and I  want to assert that no matters the value, none of those methods are called. 

For this case `any()` is the one that suits what it's being tested.

File: rskj-core/src/main/java/co/rsk/peg/BridgeStorageProvider.java (hunk 18 above)
Comment (nathanieliov) [reply rskj-core/src/main/java/co/rsk/peg/BridgeStorageProvider.java:940]: Done

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java (hunk 17 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java]: Updated

File: rskj-core/src/main/java/co/rsk/peg/BridgeStorageIndexKey.java (hunk 16 above)
Comment (nathanieliov) [reply rskj-core/src/main/java/co/rsk/peg/BridgeStorageIndexKey.java]: Updated

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java (hunk 15 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java]: Updated

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java (hunk 14 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java]: Done

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java (hunk 13 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java]: Updated

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java (hunk 11 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderBridgeTxsIndexTests.java]: Done

File: rskj-core/src/main/java/co/rsk/peg/BridgeStorageProvider.java (hunk 10 above)
Comment (nathanieliov) [reply rskj-core/src/main/java/co/rsk/peg/BridgeStorageProvider.java]: Updated

File: rskj-core/src/main/java/co/rsk/peg/BridgeStorageProvider.java [hunk 19]
```diff
@@ -923,6 +925,51 @@ protected int getReleaseRequestQueueSize() throws IOException {
...
+        if (!activations.isActive(RSKIP379) || sigHash == null) {
+            return;
+        }
+
+        if (hasBridgeBtcTxSigHash(sigHash)){
+            throw new IllegalStateException(String.format("Given bridge btc tx sigHash %s already exists in the index. Index entries are considered unique.", sigHash));
+        }
+
+        if (bridgeBtcTxSigHashes == null){
+            bridgeBtcTxSigHashes = new HashSet<>();
```
Comment (kelvinator07) [inline rskj-core/src/main/java/co/rsk/peg/BridgeStorageProvider.java:953]: We Keep Adding To This And It Keeps Growing

File: rskj-core/src/main/java/co/rsk/peg/BridgeStorageProvider.java [hunk 20]
```diff
@@ -923,6 +925,51 @@ protected int getReleaseRequestQueueSize() throws IOException {
...
+
+        bridgeBtcTxSigHashes.add(sigHash);
+    }
+
+    private void saveBridgeBtcTxSigHashes() {
+        if (!activations.isActive(RSKIP379) || bridgeBtcTxSigHashes == null) {
+            return;
+        }
+
+        bridgeBtcTxSigHashes.forEach(bridgeBtcTxSigHash -> repository.addStorageBytes(
```
Comment (kelvinator07) [inline rskj-core/src/main/java/co/rsk/peg/BridgeStorageProvider.java:964]: Why Do We Try To Save The Same hashes Existing Already In **bridgeBtcTxSigHashes** Everytime We Call **saveBridgeBtcTxSigHashes**
Shouldn't That Throw Like An Error Or Something?

Comment (kelvinator07) [inline rskj-core/src/main/java/co/rsk/peg/BridgeStorageProvider.java:964]:
```suggestion
       log.warn("[saveBridgeBtcTxSigHashes] About to save {} sigh hashes.", bridgeBtcTxSigHashes.size())
        bridgeBtcTxSigHashes.forEach(bridgeBtcTxSigHash -> repository.addStorageBytes(
```
Nice To Have a Log Here Incase We Encounter The Rare Case We Talked About

Comment (nathanieliov) [reply rskj-core/src/main/java/co/rsk/peg/BridgeStorageProvider.java]: Can you remind me what case it is? Even though I believe no log is needed. There is not even a logger instance currently in the bridge storage provider class. None of the storage provider methods is logging anything now.

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderPegoutTxIndexTests.java [hunk 21]
```diff
@@ -0,0 +1,511 @@
...
+        return Stream.of(
+            Arguments.of(new byte[]{1, 2}),
+            Arguments.of(new byte[]{0})
+        );
+    }
+
+    private static Stream<Arguments> valid_sigHash_parameters() {
+        return Stream.of(
+            Arguments.of(false, Sha256Hash.ZERO_HASH),
+            Arguments.of(false, PegTestUtils.createHash(10)),
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderPegoutTxIndexTests.java:53]: Let's add this function to BitcoinTestUtils and start using it from there. We can still leave the one in PegTestUtils for the moment since it's probably used in many places, but for this class and new tests let's use the one in BitcoinTestUtils

**Review by marcos-iov — COMMENTED** [review]
Well done! Just a minor comment

File: rskj-core/src/main/java/co/rsk/peg/BridgeStorageProvider.java (hunk 20 above)
Comment (kelvinator07) [reply rskj-core/src/main/java/co/rsk/peg/BridgeStorageProvider.java]: The Case Where We Have Both a Pegout And Migration Transaction At The Same Time. So We Will Have 2 Hashes Set To `bridgeBtcTxSigHashes`. Its a Rare Case That We Want To Log If It Happens For Debugging Sake.

File: rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderPegoutTxIndexTests.java (hunk 21 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeStorageProviderPegoutTxIndexTests.java]: Done

File: rskj-core/src/main/java/co/rsk/peg/BridgeStorageProvider.java (hunk 20 above)
Comment (nathanieliov) [reply rskj-core/src/main/java/co/rsk/peg/BridgeStorageProvider.java]: If that happens, it will be fine. I changed the implementation of the index. Now it supports a list of `pegoutTxSigHashes`, so it is capable of storing `n` sigHashes per rsk tx as long as they are different. 

However, the case you mention will only be possible if we store the entries at the moment when peg-outs are created. But we do not pretend to do it at that time. We will do it once the peg-out is fully signed. This is on the `addSignature` method. If in the future the bridge adds the capability to add a signature for more than one peg-out at a time, no prob, we are covered.

So basically, we are fine so far.

**Review by marcos-iov — APPROVED** [review]
